fix: miscellaneous conversion of Provider<File> to RegularFileProvide…

DPUkyle opened the pull request:
…r and DirectoryProvider

 - This enables configuration of the tasks via Java, avoiding `SpotBugsTask#reports(Closure)`
 - Also add `@Inject` to SpotBugsTask constructor per https://guides.gradle.org/using-the-worker-api/#converting_to_the_worker_api

DPUkyle committed:
fix #53: miscellaneous conversion of Provider<File> to RegularFileProvider and DirectoryProvider

 - This enables configuration of the tasks via Java, avoiding `SpotBugsTask#reports(Closure)`
 - Also add @Inject to SpotBugsTask constructor per https://guides.gradle.org/using-the-worker-api/#converting_to_the_worker_api

KengoTODA reviewed (changes requested):
Seems nice, thank you. I found one problem around `$buildDir` handling, then created a patch to reproduce:
[patches.zip](https://github.com/KengoTODA/spotbugs-gradle-plugin-v2/files/4069370/patches.zip)

I also proposed a fix, please check this ZIP file.

DPUkyle commented:
Thank you for the patches; I will apply and update the PR tomorrow, 2020-01-16 US PT.

DPUkyle committed:
fix: apply Kengo's patch and test for buildDir customization and lazy report instantiation

DPUkyle committed:
Add multiversion testing; 6.1 fails

  - For context, see these issues:
  - https://github.com/spotbugs/spotbugs-gradle-plugin/pull/174
  - https://github.com/gradle/gradle/issues/11189
  - https://github.com/gradle/gradle/issues/11923

DPUkyle commented:
I knew one test fails for Gradle 6.1; I'm asking for help from one of the Gradle engineers.

DPUkyle committed:
fix: Restore use of RegularFileProperty

KengoTODA commented:
Still have one problem:

> Some problems were found with the configuration of task ':spotbugsMain' (type 'SpotBugsTaskForJava').
> > No value has been specified for property 'firstEnabledReport.outputLocation'.
> > No value has been specified for property 'firstEnabledReport.required'.

I checked both of gradle/gradle#11189 and gradle/gradle#11923
it's strange that we face this problem even though we depend on no internal API 🤔

DPUkyle commented:
Yes I agree it is strange; it is related to new APIs added in Gradle 6.1. I'm asking a Gradle engineer for help offline. I will be in touch. Thank you again for your patience and collaboration :bow:.

KengoTODA commented:
Thank you.

After you found a solution and commit it, please run `./gradlew spotlessApply` or `./gradlew` to reformat your code.

KengoTODA commented:
#60 introduced a mechanism to test with multiple gradle versions, please use it if possible.

DPUkyle commented:
Great, I will rebase and use it once I have a fix.

DPUkyle committed:
style: spotless fixes

DPUkyle committed:
fix: Enable Gradle 6.1 compatibility; add to test matrix

 - minor: fix system property default value for local testing

KengoTODA commented:
So that Gradle gave up supporting backward compatibility? What's the difference btw enabled and required?

DPUkyle commented:
> So that Gradle gave up supporting backward compatibility? 

I understand your frustration and I am not happy either - this was more difficult than it should be. In this case, directly instantiating the report classes was incorrect, instead objectFactory should be used. Gradle will apparently decorate the instance with missing methods. But how could we know this as it is not documented AFAIK. Further, while the Report API is a public interface I'm not sure any external plugins have successfully implemented it. Spotbugs v2 will be the best example 😄 

> What's the difference btw enabled and required?

I wish I knew. I think this additive change in a minor release is a mistake. Specifically, I believe the "decorate" functionality I mentioned above did not anticipate `getFirstEnabledReport()` is annotated with `@Nested`. I do not understand the intent of the new field.

As plugin authors I think we have no choice but to continue to put a strong emphasis on multi-version testing to protect ourselves against and adapt to the frequently-changing APIs. Also, if we can get this plugin added to Gradle's third-party suite (see #59), this might help them better understand the impact of their changes, and help us keep the plugin implementation more correct.

KengoTODA reviewed (approved):
Thanks for your great contribution!

To make release note simple and easy to understand, I will squash-and-merge these changes.
